### ISSUE TITLE
"goimports" not run on save after a while

### ISSUE THREAD

**Issue by atombender** [microsoft/vscode-go#2599]
My config:

```json
{
  "go.formatTool": "goimports",
  "[go]": {
    "editor.formatOnSave": true,
    "editor.codeActionsOnSave": {
      "source.organizeImports": true
    },
    "editor.codeActionsOnSaveTimeout": 5000
  },
  "go.useLanguageServer": true,
  "go.languageServerExperimentalFeatures": {
    "format": false
  },
  "go.languageServerFlags": [
    "-trace"
  ]
}
```

As per [this Gopls issue](https://github.com/golang/go/issues/31977), I have modified Gopls to _not_ format imports, because it has some severe performance regressions where the import-resolving logic takes anywhere between 20 and 50 seconds to run. Until that's fixed, I want to rely entirely on `goimports`.

(Note, in case it's relevant, that in that issue, I was advised to set `"format": false"` under `go.languageServerExperimentalFeatures`, but that setting seems not to be honoured at all; while Gopls has a command-line option to disable formatting, [in one relevant](https://github.com/golang/tools/blob/8ea4f8e3e5bfd414380c369ad66f924066e789f7/internal/lsp/source/format.go#L67) place it's not set. I haven't looked at the VSCode extension code to see what's going on.)

Anyway, in my case I'm programming as usual and Gopls seems happy, until suddenly, `goimports` in longer invoked when I save a file. My uneducated guess is that some state related to Gopls turns it off. I _don't_ see the _Running save participants_ status in the status bar. Reloading the window is the only thing that helps. I've tried waiting a while, and it doesn't help. Killing `gopls` does not help.

I've tried tailing all of VSCode's log files, and don't see any activity or any errors. I do see `golangci-lint`, my configured linter, being run.

**Comment by lloiser** [comment]
Have a look at this thread: https://github.com/microsoft/vscode-go/issues/2336

**tl;dr**: increase the `editor.formatOnSaveTimeout` setting

**Comment by lloiser** [comment]
Or checkout the [FAQ](https://github.com/Microsoft/vscode-go/wiki/Go-with-VS-Code-FAQ-and-Troubleshooting) and have a look at the _Q: Why doesn't formatting doesn't work on file save?_ question.

**Comment by atombender** [comment]
Did you read the part about how `goimports` isn't being invoked at all and the "Running save participants" status isn't showing?

I'm monitoring processes while saving and `goimports` doesn't seem to be spawned at all. The timeout should be irrelevant here.

**Comment by ramya-rao-a** [comment]
@atombender 

In the below section, can set `"source.organizeImports"` to `false`? This will ensure that gopls will not try find missing imports or re-order your existing imports on each file save. This is just eliminate any edit conflicts between goimports and gopls until we figure out what is going on

```
"[go]": {
    "editor.formatOnSave": true,
    "editor.codeActionsOnSave": {
      "source.organizeImports": true
    },
    "editor.codeActionsOnSaveTimeout": 5000
  }
```

> My uneducated guess is that some state related to Gopls turns it off.

By setting `format` to `false` in `go.languageServerExperimentalFeatures`, what is happening is that the Go extension [skips requesting the language server for any formatting changes](https://github.com/microsoft/vscode-go/blob/0.11.0/src/goMain.ts#L144-L149) and runs its own formatting provider which runs the tool as specified in `go.formatTool` setting. `gopls` has no say in formatting, once the `format` in `go.languageServerExperimentalFeatures` and `source.organizeImports` is set to false


The next time this happens, can you check if manual formatting still works? Right-click -> Format Document?

**Comment by atombender** [comment]
Still happening. I have a window open now where save does not trigger. Some observations:

1. `Format document` action complains about there being no default formatter.
2. If I go to configure it, _two_ formatters are available:

![Screen Shot 2019-06-26 at 21 18 14](https://user-images.githubusercontent.com/50314/60226150-ef8bbd00-9857-11e9-8cb4-c08d9b0cae19.jpg)

3. No matter which one I set to be the default, `Format document` still doesn't work.
4. If I do `Format document with...`, the list now shows both "Go" formatters are being the default:

![Screen Shot 2019-06-26 at 20 36 49](https://user-images.githubusercontent.com/50314/60224880-6c1b9d00-9852-11e9-8078-49e30049b520.jpg)

5. If I do `Format document with...` and select the _first_ formatter, nothing happens.
6. If I do `Format document with...` and select the _second_ formatter, it formats the code, and I can see that it's spawning `goimports`.

If I reload the window, things are back to normal; formatting happens on save and `Format document` works. _But_, if I do `Format document with...` I can see that the first formatter in the list does nothing, and the second formatter works.

Configuring the formatter puts this in my `settings.json` no matter which one I choose; i.e. both have the same ID:

```
  "[go]": {
    // ...
    "editor.defaultFormatter": "ms-vscode.Go"
  }
```

What this feels like to me: There are two formatters being registered with the same ID. Only one of them works. Somehow, VSCode usually picks the right one, but after a while, it starts picking the wrong one. Why this isn't completely random, I don't know, but maybe there's a cache somewhere that has a dictionary of ID -> formatter?

Any explanation for why there are two identically named formatters being offered?

**Comment by atombender** [comment]
Looking at the logs, I do see lines like this that I didn't see when tailing before:

```
[2019-06-25 23:40:17.177] [exthost] [error] [ms-vscode.Go] provider FAILED
```

No more detail than that.

I'm also seeing some older entries like this, before I disabled Gopls's formatting, are these anything bad?

```
[2019-06-23 21:38:16.667] [exthost] [warning] ms-vscode.Go -Code actions of kind 'source.organizeImports 'requested but returned code action is of kind 'quickfix'. Code action will be dropped. Please check 'CodeActionContext.only' to only return requested code actions.
```

**Comment by atombender** [comment]
I'm seeing more `exthost` log output right now that wasn't showing before for some reason. Related?

```
[2019-06-28 12:48:24.775] [exthost] [error] [ms-vscode.Go] provider FAILED
[2019-06-28 12:48:24.775] [exthost] [error] Error: selectionRange must be contained in fullRange
	at Function.validate (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:345:152)
	at new w (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:345:20)
	at new w (/Applications/Visual Studio Code.app/Contents/Resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:333:525)
	at asDocumentSymbol (/Users/alex/.vscode/extensions/ms-vscode.go-0.11.0/node_modules/vscode-languageclient/lib/protocolConverter.js:360:22)
	at Array.map (<anonymous>)
	at Object.asDocumentSymbols (/Users/alex/.vscode/extensions/ms-vscode.go-0.11.0/node_modules/vscode-languageclient/lib/protocolConverter.js:357:23)
	at registerLanguageProvider.client.sendRequest.then (/Users/alex/.vscode/extensions/ms-vscode.go-0.11.0/node_modules/vscode-languageclient/lib/client.js:983:62)
```

**Comment by ramya-rao-a** [comment]
Thanks for all the details @atombender 
https://github.com/microsoft/vscode-go/issues/2446 explains how there can be 2 formatters that can get registered.

Until we get that resolved, you can try to use the formatting feature from the language server.

**Comment by atombender** [comment]
Thanks. Why does registering it twice cause it to stop working, though?

Unfortunately, I can't use Gopls for formatting. It's pretty broken these days; I'm sigkilling it every 10 minutes or so because it ends up in a bad state where autocompletions don't work and it reports error regions for things that have been fixed and so on. It's painful, and one part of it which is particularly broken is its formatting/import reorganization, as I mentioned in the issue. It's unusable. I have to rely on `goimports`.

**Comment by ramya-rao-a** [comment]
> Why does registering it twice cause it to stop working, though?

Maybe because each attempt at formatting tries to fetch the formatter and it is a game of chance which of the 2 (the one from the extension, the other from the language server) gets picked up? Not sure...

@jrieken Does each formatting request try to figure out which provider to use? Or does it pick one for a VS Code session and stick with it? @atombender is seeing the case of https://github.com/microsoft/vscode/issues/72315 and has the format feature working fine for a while and then stop mid session.

**Comment by atombender** [comment]
Ah, so one of them is Gopls. Is there a way for me to comment out the Gopls registration somewhere?

**Comment by ramya-rao-a** [comment]
Based on https://github.com/microsoft/vscode/issues/70314#issuecomment-472336416 it looks like that should happen on the language server side

Let's continue this on https://github.com/microsoft/vscode-go/issues/2446#issuecomment-506917579
